manager/orchestrator/constraint_enforcer.go

```diff
@@ -0,0 +1,137 @@
+package orchestrator
+
+import (
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/constraint"
+	"github.com/docker/swarmkit/manager/state"
+	"github.com/docker/swarmkit/manager/state/store"
+)
+
+// ConstraintEnforcer watches for updates to nodes and shuts down tasks that no
+// longer satisfy scheduling constraints or resource limits.
+type ConstraintEnforcer struct {
+	store       *store.MemoryStore
+	watcher     chan events.Event
+	cancelWatch func()
+	stopChan    chan struct{}
+	doneChan    chan struct{}
+}
+
+// NewConstraintEnforcer creates a new ConstraintEnforcer.
+func NewConstraintEnforcer(store *store.MemoryStore) *ConstraintEnforcer {
+	watcher, cancel := state.Watch(store.WatchQueue(), state.EventUpdateNode{})
+
+	return &ConstraintEnforcer{
+		store:       store,
+		watcher:     watcher,
+		cancelWatch: cancel,
+		stopChan:    make(chan struct{}),
+		doneChan:    make(chan struct{}),
+	}
+}
+
+// Run is the ConstraintEnforcer's main loop.
+func (ce *ConstraintEnforcer) Run() {
+	defer close(ce.doneChan)
+
+	for {
+		select {
+		case event := <-ce.watcher:
+			node := event.(state.EventUpdateNode).Node
+
+			var (
+				tasks []*api.Task
+				err   error
+			)
+
+			ce.store.View(func(tx store.ReadTx) {
+				tasks, err = store.FindTasks(tx, store.ByNodeID(node.ID))
+			})
+
+			if err != nil {
+				log.L.WithError(err).Errorf("failed to list tasks for node ID %s", node.ID)
+			}
+
+			var availableMemoryBytes, availableNanoCPUs int64
+			if node.Description != nil && node.Description.Resources != nil {
+				availableMemoryBytes = node.Description.Resources.MemoryBytes
+				availableNanoCPUs = node.Description.Resources.NanoCPUs
+			}
+
+			removeTasks := make(map[string]*api.Task)
+
+			// TODO(aaronl): The set of tasks removed will be
+			// nondeterministic because it depends on the order of
+			// the slice returned from FindTasks. We could do
+			// a separate pass over the tasks for each type of
+			// resource, and sort by the size of the reservation
+			// to remove the most resource-intensive tasks.
```

Comment:
On a related topic, we may need to define service/task priority so that some tasks should have better chance of running. For example, system services (likely global services) including monitoring should be run on every node when possible.